network/src/noise/handshake.rs

```diff
@@ -298,6 +298,19 @@ impl NoiseUpgrader {
             )
         })?;
 
+        // prevent accidental self-dials
+        // this situation could occur either as a result of our own discovery
+        // mis-configuration or a potentially malicious discovery peer advertising
+        // a (loopback ip or mirror proxy) and our public key.
+        if remote_peer_id == self.self_peer_id {
+            // TODO(philiphayes): security logging. someone should investigate
+            // on-chain reconfiguration history to see if someone is misbehaving.
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "noise: detected accidental self-dial: we have the same peer id as the client",
+            ));
```

Comment:
I was thinking maybe we want to put `peer_id` in the error, but then I realized that may be added in logging in the future.  Thoughts?

Reply:
self_peer_id should get logged w/ the NetworkContext